### PR TITLE
Refactor gcp cluster instance

### PR DESCRIPTION
**Pull request by praveenramuu**
- refactor gcp instance to move count to resource level
 - add default to null for input variables and change variable names

### CODE REVIEW

File: resources/gcp/compute/vm_instance/vm_instance.tf [hunk 1]
```diff
@@ -8,12 +8,6 @@ variable "zone" {
   description = "GCP zone that the instances should be created."
 }
```
Comment (sasikeda) [inline resources/gcp/compute/vm_instance/vm_instance.tf]: I think zone needs to be delete

File: resources/gcp/compute/vm_instance/vm_instance.tf [hunk 2]
```diff
@@ -130,26 +124,55 @@ variable "public_key_content" {
...
+  type        = number
+  default     = 0
+  description = "Number of Instance that needs to create."
+}
+
+locals {
+  vm_configuration    = flatten([for zone_name in var.vpc_availability_zones : [for network in var.vpc_subnets : [for i in range(var.total_cluster_instances) : {zone = zone_name , subnet = network, vm_name = "${var.instance_name}-${i}"}]]])
+  disk_configuration  = flatten([for vm_index in range(var.total_cluster_instances) : [for disk_no in range(var.total_persistent_disks) : { index = vm_index , vm_name_suffix = "${disk_no}" }]])
+
+  block_device_names = ["/dev/sdb", "/dev/sdc", "/dev/sdd", "/dev/sdf", "/dev/sdg",
```
Comment (sasikeda) [inline resources/gcp/compute/vm_instance/vm_instance.tf:151]: Ideally, 256 attachments are supported. But we can stick to above.

File: resources/gcp/compute/vm_instance/vm_instance.tf [hunk 3]
```diff
@@ -130,26 +124,55 @@ variable "public_key_content" {
...
 resource "google_compute_instance" "scale_instance" {
-  name         = var.instance_name
+  count        = length(local.vm_configuration)
+  name         = "${local.vm_configuration[count.index].vm_name}"
   machine_type = var.machine_type
-  zone         = var.zone
+  zone         = local.vm_configuration[count.index].zone
 
   allow_stopping_for_update = true
   tags                      = var.vm_instance_tags
```
Comment (sasikeda) [inline resources/gcp/compute/vm_instance/vm_instance.tf:178]: No need of tags

File: resources/gcp/compute/vm_instance/vm_instance.tf [hunk 4]
```diff
@@ -8,12 +8,6 @@ variable "zone" {
...
 
-variable "subnet_name" {
-  type        = string
-  nullable    = true
-  description = "Instance subnet name."
-}
-
 variable "operator_email" {
   type        = string
   description = "GCP service account e-mail address."
```
Comment (sasikeda) [inline resources/gcp/compute/vm_instance/vm_instance.tf:7]: rename operator_email to service_email

File: resources/gcp/compute/vm_instance/vm_instance.tf [hunk 5]
```diff
@@ -167,7 +190,7 @@ resource "google_compute_instance" "scale_instance" {
   }
 
   network_interface {
-    subnetwork = var.subnet_name
+    subnetwork = local.vm_configuration[count.index].subnet
```
Comment (sasikeda) [inline resources/gcp/compute/vm_instance/vm_instance.tf:193]: is there a need of scratch disk loop ?

File: resources/gcp/compute/vm_instance/vm_instance.tf [hunk 6]
```diff
@@ -208,23 +232,24 @@ resource "google_compute_disk" "data_disk" {
 }
 
 resource "google_compute_attached_disk" "attach_data_disk" {
-  count      = length(google_compute_disk.data_disk)
-  disk       = google_compute_disk.data_disk[count.index].id
-  instance   = google_compute_instance.scale_instance.id
+  count      = length(local.disk_configuration)
+  zone       = google_compute_instance.scale_instance[local.disk_configuration[count.index].index].zone
+  disk       = format("%s-%s-%s", var.data_disk_name_prefix, google_compute_instance.scale_instance[local.disk_configuration[count.index].index].instance_id, local.disk_configuration[count.index].vm_name_suffix)
+  instance   = google_compute_instance.scale_instance[local.disk_configuration[count.index].index].instance_id
```
Comment (sasikeda) [inline resources/gcp/compute/vm_instance/vm_instance.tf]: instance_id ?, from doc it suggests name or self_link
https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_attached_disk

File: resources/gcp/compute/vm_instance/vm_instance.tf [hunk 7]
```diff
@@ -208,23 +232,24 @@ resource "google_compute_disk" "data_disk" {
...
   depends_on = [google_compute_disk.data_disk]
 }
 
 #Instance details
-output "scale_instance_ids" {
-  value = google_compute_instance.scale_instance.instance_id
+output "instance_ids" {
+  value = google_compute_instance.scale_instance[*].instance_id
```
Comment (sasikeda) [inline resources/gcp/compute/vm_instance/vm_instance.tf]: kindly use self_link

File: resources/gcp/compute/vm_instance/vm_instance.tf [hunk 8]
```diff
@@ -241,15 +266,16 @@ output "data_disk_attachment_id" {
 }
 
 output "subnet_name" {
-  value = var.subnet_name
+  value = google_compute_disk.data_disk[*].zone
```
Comment (sasikeda) [inline resources/gcp/compute/vm_instance/vm_instance.tf:268]: subnet_name --> but o/p is showing zone ?

File: gcp_scale_templates/sub_modules/instance_template/main.tf [hunk 9]
```diff
@@ -44,6 +44,9 @@ locals {
 
   gpfs_base_rpm_path = var.spectrumscale_rpms_path != null ? fileset(var.spectrumscale_rpms_path, "gpfs.base-*") : null
   scale_version      = local.gpfs_base_rpm_path != null ? regex("gpfs.base-(.*).x86_64.rpm", tolist(local.gpfs_base_rpm_path)[0])[0] : null
+
+  total_local_ssd_disks  = var.data_disk_type == "local-ssd" ? var.block_devices_per_storage_instance : 0
```
Comment (sasikeda) [inline gcp_scale_templates/sub_modules/instance_template/main.tf]: kindly use seperate variables for local_ssd_disks and persistent_disks, as from cloudkit we control then via different variables.

File: gcp_scale_templates/sub_modules/instance_template/main.tf [hunk 10]
```diff
@@ -136,50 +139,49 @@ module "compute_cluster_instances" {
...
-  total_data_disks              = var.block_devices_per_storage_instance
-  instance_name_prefix          = format("%s-storage-tie", var.resource_prefix)
-  machine_type                  = var.compute_machine_type
-  subnet_name                   = var.vpc_storage_cluster_private_subnets != null ? length(var.vpc_storage_cluster_private_subnets) > 1 ? [var.vpc_storage_cluster_private_subnets[2]] : [] : []
+  source                        = "../../../resources/gcp/compute/vm_instance"
+  vpc_availability_zones        = var.vpc_availability_zones
+  ssh_key_path                  = var.storage_cluster_public_key_path
+  ssh_user_name                 = var.instances_ssh_user_name
+  total_cluster_instances       = var.vpc_storage_cluster_private_subnets != null ? ((length(var.vpc_storage_cluster_private_subnets) > 1 && (local.cluster_type == "storage" || local.cluster_type == "combined")) ? 1 : 0) : 0
+  total_persistent_disks        = 5
```
Comment (sasikeda) [inline gcp_scale_templates/sub_modules/instance_template/main.tf]: we are creating 5 persistent disks here ?
it should be 1 disk with 5GB size

File: gcp_scale_templates/sub_modules/instance_template/main.tf [hunk 11]
```diff
@@ -136,50 +139,49 @@ module "compute_cluster_instances" {
...
-  zone                          = var.vpc_availability_zones[2]
-  instances_ssh_public_key_path = var.storage_cluster_public_key_path
-  instances_ssh_user_name       = var.instances_ssh_user_name
-  total_cluster_instances       = 1
-  total_data_disks              = var.block_devices_per_storage_instance
-  instance_name_prefix          = format("%s-storage-tie", var.resource_prefix)
-  machine_type                  = var.compute_machine_type
-  subnet_name                   = var.vpc_storage_cluster_private_subnets != null ? length(var.vpc_storage_cluster_private_subnets) > 1 ? [var.vpc_storage_cluster_private_subnets[2]] : [] : []
+  source                        = "../../../resources/gcp/compute/vm_instance"
+  vpc_availability_zones        = var.vpc_availability_zones
```
Comment (sasikeda) [inline gcp_scale_templates/sub_modules/instance_template/main.tf]: it will be placed in 3rd az. not all vpc zones needs to be provided.

**Review by sasikeda — CHANGES_REQUESTED** [review]

File: resources/gcp/compute/vm_instance/vm_instance.tf (hunk 1 above)
Comment (praveenramuu) [reply resources/gcp/compute/vm_instance/vm_instance.tf]: Thanks sasi , done

File: resources/gcp/compute/vm_instance/vm_instance.tf (hunk 2 above)
Comment (praveenramuu) [reply resources/gcp/compute/vm_instance/vm_instance.tf:151]: Thanks sasi

File: resources/gcp/compute/vm_instance/vm_instance.tf (hunk 4 above)
Comment (praveenramuu) [reply resources/gcp/compute/vm_instance/vm_instance.tf:7]: Thanks done

File: resources/gcp/compute/vm_instance/vm_instance.tf (hunk 5 above)
Comment (praveenramuu) [reply resources/gcp/compute/vm_instance/vm_instance.tf:193]: Thanks , It's already exists at line 203 with dynamic "scratch_disk" block

File: resources/gcp/compute/vm_instance/vm_instance.tf (hunk 7 above)
Comment (praveenramuu) [reply resources/gcp/compute/vm_instance/vm_instance.tf]: Thanks sasi , done

File: resources/gcp/compute/vm_instance/vm_instance.tf (hunk 8 above)
Comment (praveenramuu) [reply resources/gcp/compute/vm_instance/vm_instance.tf:268]: Thanks sasi , changed to data_disk_zone :-)

File: resources/gcp/compute/vm_instance/vm_instance.tf (hunk 6 above)
Comment (praveenramuu) [reply resources/gcp/compute/vm_instance/vm_instance.tf]: Thanks sasi , both works name as well as instance_id .
instance_id is used for instance param in above mentioned doc at 'example usage'

File: gcp_scale_templates/sub_modules/instance_template/main.tf (hunk 9 above)
Comment (praveenramuu) [reply gcp_scale_templates/sub_modules/instance_template/main.tf]: Thanks sasi , done , introduced block_devices_per_storage_instance & scratch_devices_per_storage_instance

File: gcp_scale_templates/sub_modules/instance_template/main.tf (hunk 10 above)
Comment (praveenramuu) [reply gcp_scale_templates/sub_modules/instance_template/main.tf]: Thanks done

File: gcp_scale_templates/sub_modules/instance_template/main.tf (hunk 11 above)
Comment (praveenramuu) [reply gcp_scale_templates/sub_modules/instance_template/main.tf]: Thanks done sasi

File: resources/gcp/compute/vm_instance/vm_instance.tf (hunk 3 above)
Comment (praveenramuu) [reply resources/gcp/compute/vm_instance/vm_instance.tf:178]: I will raise new PR for this removal of tag

**Comment by sasikeda** [comment]
Kindly change main.tf (lines), inoder to work with null.

```
locals {
  cluster_type = (
    (length(var.vpc_storage_cluster_private_subnets) > 0 && length(var.vpc_compute_cluster_private_subnets) == 0) ? "storage" :
    (length(var.vpc_storage_cluster_private_subnets) == 0 && length(var.vpc_compute_cluster_private_subnets) > 0) ? "compute" :
    (length(var.vpc_storage_cluster_private_subnets) > 0 && length(var.vpc_compute_cluster_private_subnets) > 0) ? "combined" : "none"
  )
```
to

```
cluster_type = (
    (var.vpc_storage_cluster_private_subnets != null && var.vpc_compute_cluster_private_subnets == null) ? "storage" :
    (var.vpc_storage_cluster_private_subnets == null && var.vpc_compute_cluster_private_subnets != null) ? "compute" :
    (var.vpc_storage_cluster_private_subnets != null && var.vpc_compute_cluster_private_subnets != null) ? "combined" : "none"
  )
```

**Comment by sasikeda** [comment]
The conditions defined for vpc_subnets in the compute only, storage only, tie-breaker modules break if we choose to deploy only compute or only storage;
```
Error: Iteration over null value
│
│   on ../../../resources/gcp/compute/vm_instance/vm_instance.tf line 142, in locals:
│  142:   vm_configuration    = flatten([for zone_name in var.vpc_availability_zones : [for network in var.vpc_subnets : [for i in range(var.total_cluster_instances) : {zone = zone_name , subnet = network, vm_name = "${var.instance_name}-${i}"}]]])
│     ├────────────────
│     │ var.vpc_subnets is null
│
│ A null value cannot be used as the collection in a 'for' expression.
```

**Comment by praveenramuu** [comment]
> The conditions defined for vpc_subnets in the compute only, storage only, tie-breaker modules break if we choose to deploy only compute or only storage;
> 
> ```
> Error: Iteration over null value
> │
> │   on ../../../resources/gcp/compute/vm_instance/vm_instance.tf line 142, in locals:
> │  142:   vm_configuration    = flatten([for zone_name in var.vpc_availability_zones : [for network in var.vpc_subnets : [for i in range(var.total_cluster_instances) : {zone = zone_name , subnet = network, vm_name = "${var.instance_name}-${i}"}]]])
> │     ├────────────────
> │     │ var.vpc_subnets is null
> │
> │ A null value cannot be used as the collection in a 'for' expression.
> ```
Thanks sasi , it works if we pass empty array i.e, [] for inputs vpc_storage_cluster_private_subnets/vpc_compute_cluster_private_subnets. but yes for null value it fails which I have not unit tested  for this scenario.  Added code now to handle the null for input variables.

**Comment by praveenramuu** [comment]
> Kindly change main.tf (lines), inoder to work with null.
> 
> ```
> locals {
>   cluster_type = (
>     (length(var.vpc_storage_cluster_private_subnets) > 0 && length(var.vpc_compute_cluster_private_subnets) == 0) ? "storage" :
>     (length(var.vpc_storage_cluster_private_subnets) == 0 && length(var.vpc_compute_cluster_private_subnets) > 0) ? "compute" :
>     (length(var.vpc_storage_cluster_private_subnets) > 0 && length(var.vpc_compute_cluster_private_subnets) > 0) ? "combined" : "none"
>   )
> ```
> 
> to
> 
> ```
> cluster_type = (
>     (var.vpc_storage_cluster_private_subnets != null && var.vpc_compute_cluster_private_subnets == null) ? "storage" :
>     (var.vpc_storage_cluster_private_subnets == null && var.vpc_compute_cluster_private_subnets != null) ? "compute" :
>     (var.vpc_storage_cluster_private_subnets != null && var.vpc_compute_cluster_private_subnets != null) ? "combined" : "none"
>   )
> ```

Thanks changed sasi

**Comment by sasikeda** [comment]
Looks the new changes, is causing count to multiple. Kindly request to verify, maybe my test could be incorrect.

**Review by sasikeda — APPROVED** [review]

**Pull request merged**